File: dotnet/Trinsic/CredentialsService.cs
using System;
using System.Threading.Tasks;
using Trinsic.Sdk.Options.V1;
using Trinsic.Services.Common.V1;
using Trinsic.Services.VerifiableCredentials.Templates.V1;
using Trinsic.Services.VerifiableCredentials.V1;

namespace Trinsic;

public class CredentialsService : ServiceBase
{
    public CredentialsService(ServiceOptions options)
        : base(options) {
        Client = new(Channel);
    }

    public CredentialsService() {
        Client = new(Channel);
    }
    
    internal CredentialsService(ITokenProvider tokenProvider) : base(new(), tokenProvider) {
        Client = new(Channel);
    }

    private VerifiableCredential.VerifiableCredentialClient Client { get; }

    /// <summary>
    /// Signs an input credential
    /// </summary>
    /// <param name="request"></param>
    /// <returns></returns>
    public async Task<IssueResponse> IssueCredentialAsync(IssueRequest request) {
        if (string.IsNullOrWhiteSpace(request.DocumentJson)) {
            throw new ArgumentException("document json must not be empty");
        }
        return await Client.IssueAsync(request, await BuildMetadataAsync(request));
    }

    public IssueResponse IssueCredential(IssueRequest request) {
        if (string.IsNullOrWhiteSpace(request.DocumentJson)) {
            throw new ArgumentException("document json must not be empty");
        }
        return Client.Issue(request, BuildMetadata(request));
    }

    /// <summary>
    /// Issue a verifiable credential from a predefined <see cref="TemplateData"/>.
    /// </summary>
    /// <param name="request">The request object with the template identifier and the values</param>
    /// <returns>Verifiable credential as JSON</returns>
    public async Task<string> IssueFromTemplateAsync(IssueFromTemplateRequest request) {
        var response = await Client.IssueFromTemplateAsync(request, await BuildMetadataAsync(request));
        return response.DocumentJson;
    }

    /// <summary>
    /// Issue a verifiable credential from a predefined <see cref="TemplateData"/>.
    /// </summary>
    /// <param name="request">The request object with the template identifier and the values</param>
    /// <returns>Verifiable credential as JSON</returns>
    public string IssueFromTemplate(IssueFromTemplateRequest request) {
        var response = Client.IssueFromTemplate(request, BuildMetadata(request));
        return response.DocumentJson;
    }

    /// <summary>
    /// Create a proof from a record in the user's wallet. The record must be a valid
    /// verifiable credential and contain a signature from which a proof can be derived.
    /// </summary>
    /// <param name="request"></param>
    /// <returns></returns>
    public async Task<CreateProofResponse> CreateProofAsync(CreateProofRequest request) {
        return await Client.CreateProofAsync(request, await BuildMetadataAsync(request));
    }

    /// <summary>
    /// Create a proof from a record in the user's wallet. The record must be a valid
    /// verifiable credential and contain a signature from which a proof can be derived.
    /// </summary>
    /// <param name="request"></param>
    /// <returns></returns>
    public CreateProofResponse CreateProof(CreateProofRequest request) {
        return Client.CreateProof(request, BuildMetadata(request));
    }

    /// <summary>
    /// Verifies a proof document
    /// </summary>
    /// <param name="request"></param>
    /// <returns></returns>
    public async Task<bool> VerifyProofAsync(VerifyProofRequest request) {
        var response = await Client.VerifyProofAsync(
            request: request,
            headers: await BuildMetadataAsync(request));

        return response.IsValid;
    }

    public bool VerifyProof(VerifyProofRequest request) {
        var response = Client.VerifyProof(
            request: request,
            headers: BuildMetadata(request));

        return response.IsValid;
    }

    /// <summary>
    /// Check credential template status
    /// </summary>
    /// <param name="credentialStatusId"></param>
    /// <returns></returns>
    public async Task<CheckStatusResponse> CheckStatusAsync(string credentialStatusId) {
        CheckStatusRequest request = new() {CredentialStatusId = credentialStatusId};
        return await Client.CheckStatusAsync(request, await BuildMetadataAsync(request));
    }

    public CheckStatusResponse CheckStatus(string credentialStatusId) {
        CheckStatusRequest request = new() {CredentialStatusId = credentialStatusId};
        return Client.CheckStatus(request, BuildMetadata(request));
    }

    /// <summary>
    /// Update credential template revocation status
    /// </summary>
    /// <param name="credentialStatusId"></param>
    /// <param name="revoked"></param>
    /// <returns></returns>
    public async Task UpdateStatusAsync(string credentialStatusId, bool revoked) {
        UpdateStatusRequest request = new() {CredentialStatusId = credentialStatusId, Revoked = revoked};
        var response = await Client.UpdateStatusAsync(request, await BuildMetadataAsync(request));
        if (response.Status == ResponseStatus.Success) return;
        throw new Exception($"Status not completely updated {response.Status}");
    }

    public void UpdateStatus(string credentialStatusId, bool revoked) {
        UpdateStatusRequest request = new() {CredentialStatusId = credentialStatusId, Revoked = revoked};
        var response = Client.UpdateStatus(request, BuildMetadata(request));
        if (response.Status == ResponseStatus.Success) return;
        throw new Exception($"Status not completely updated {response.Status}");
    }


    /// <summary>
    /// Sends a document to the specified destination
    /// </summary>
    /// <param name="request"></param>
    /// <returns></returns>
    public async Task SendAsync(SendRequest request) {
        var response = await Client.SendAsync(
            request: request,
            headers: await BuildMetadataAsync(request));
    }

    public void Send(SendRequest request) {
        var response = Client.Send(
            request: request,
            headers: BuildMetadata(request));
    }
}

File: python/trinsic/credentials_service.py
import json

from trinsic.proto.sdk.options.v1 import ServiceOptions
from trinsic.proto.services.verifiablecredentials.v1 import (
    VerifiableCredentialStub,
    CheckStatusResponse,
)
from trinsic.service_base import ServiceBase, ResponseStatusException


class CredentialsService(ServiceBase):
    """Wrapper for the [Credentials Service](/reference/services/Credentials-service/)"""

    def __init__(
        self,
        *,
        server_config: ServiceOptions = None,
    ):
        """
        Initialize a connection to the server.
        Args:
            server_config: The URL of the server, or a channel which encapsulates the connection already.
        """
        super().__init__(server_config)
        self.client: VerifiableCredentialStub = self.stub_with_metadata(
            VerifiableCredentialStub
        )

    async def issue_credential(self, *, document: dict) -> dict:
        """
        [Issue a new credential](/reference/services/credentials-service/#issue-credential)
        Args:
            document: Dictionary describing the credential
        Returns:
            Dictionary with the issued credential
        """
        response = await self.client.issue(document_json=json.dumps(document))
        return json.loads(response.signed_document_json)

    async def issue_from_template(self, *, template_id: str, values_json: str) -> dict:
        """
        Issue a credential from the previously stored template.
        Args:
            template_id:
            values_json:
        Returns:
            The JSON document representation of this credential as a string
        """
        return json.loads(
            (
                await self.client.issue_from_template(
                    template_id=template_id, values_json=values_json
                )
            ).document_json
        )

    async def check_status(self, *, credential_status_id: str) -> CheckStatusResponse:
        """
        Check status of a credential
        Args:
            credential_status_id:
        Returns:
            `CheckStatusResponse`
        """
        return await self.client.check_status(credential_status_id=credential_status_id)

    async def update_status(self, *, credential_status_id: str, revoked: bool) -> None:
        """
        Update the status of a credential
        Args:
            credential_status_id:
            revoked:
        """
        response = await self.client.update_status(
            credential_status_id=credential_status_id, revoked=revoked
        )
        ResponseStatusException.assert_success(
            response.status, "update credential status"
        )

    async def create_proof(
        self, *, reveal_document: dict, item_id: str = "", document: dict = None
    ) -> dict:
        """
        [Create a proof](/reference/services/wallet-service/#create-proof)
        Args:
            document: document in the wallet that is signed
            item_id: id of document in the wallet
            reveal_document: JSONLD frame describing what data is to be disclosed.
        Returns:
            The JSONLD proof
        """
        document_json = json.dumps(document) if document else None
        return json.loads(
            (
                await self.client.create_proof(
                    reveal_document_json=json.dumps(reveal_document),
                    item_id=item_id,
                    document_json=document_json,
                )
            ).proof_document_json
        )

    async def verify_proof(self, *, proof_document: dict) -> bool:
        """
        [Verify a proof](/reference/services/wallet-service/#verify-proof)
        Args:
            proof_document: Document to verify
        Returns:
            `True` if verified, `False` if not verified
        """
        return (
            await self.client.verify_proof(
                proof_document_json=json.dumps(proof_document)
            )
        ).is_valid

    async def send(
        self,
        *,
        document: dict,
        email: str = None,
        did_uri: str = None,
        didcomm_invitation_json: str = None,
    ) -> None:
        """
        [Send the provided document to the given email](/reference/services/wallet-service/#sending-documents-using-email-as-identifier)
        Args:
            didcomm_invitation_json:
            did_uri:
            document: Document to send
            email: Email to which the document is sent
        """
        response = await self.client.send(
            email=email,
            document_json=json.dumps(document),
            did_uri=did_uri,
            didcomm_invitation_json=didcomm_invitation_json,
        )
        ResponseStatusException.assert_success(response.status, "sending credential")
